### ISSUE TITLE
Feedback from writing a source generator

### ISSUE THREAD

**Issue by canton7** [dotnet/roslyn#44929]
Feedback from writing a Source Generator
=============================

I've spent the last month writing a source generator for one of my projects, and just wanted to share my experience, in case that's useful.

This isn't a specific request to change anything, rather a collection of observations and pain points. You might decide to take input from these, or not. Please feel free to close this issue at any point.

## Background

I've got a library which wraps HttpClient, making it a bit more terse to use. Users define an interface which might look a bit like this:

```cs
public interface ISomeApi
{
    [Get("users")]
    Task<User> GetUser([Query] string userId);
}
```

and call it like this:

```cs
ISomeApi api = RestClient.For<ISomeApi>("https://api.example.com");
var user = await api.GetUser("canton7");
```

Behind the scenes, I generate an implementation which looks something like this (simplified):

```cs
public class SomeApi : ISomeApi
{
    
    private readonly IRequester requester;
    public SomeApi(IRequester requester) => this.requester = requester;
    public Task<User> GetUser(string userId)
    {
        var requestInfo = new RequestInfo(HttpMethod.Get, "users");
        requestInfo.AddQueryParameter("userId", userId);
        return this.requester.RequestAsyc<User>(requestInfo);
    }
}
```

Types like `RestClient`, `IRequester` and `RequestInfo` are contained in an assembly which the user references

Generic interfaces, generic methods (with and without type constraints), interface inheritance, etc, are all supported.

I've traditionally done this generation at runtime using System.Reflection.Emit. Source generators seemed like the perfect tool, so I jumped on them. I wanted to keep the same API though, so I avoided partial classes, and instead generate an assembly-level attribute linking together a user's interface with its generated implementation, then instantiate the implementation using reflection in `RestClient.For<T>`.

## Turning symbols into strings

Probably the trickiest part was generating the type and method signatures as strings. I'm lucky that I had the interface being implemented as an `INamedTypeSymbol`, and so I could use `ISymbol.ToDisplayString` (and [this long list of `SymbolDisplayFormats`](https://github.com/canton7/RestEase/blob/develop/src/RestEase.SourceGenerator/Implementation/SymbolDisplayFormats.cs)) to get me most of the way there. This has logic to handle things like:

1. Fully-qualifying types (including `global::` and the full namespace)
2. Escaping keywords, like `event` or `params`
3. Generating generic type constraints
4. Generating method declarations, including default values, `in`/`out`/`ref, etc

However, I always wondered whether I was using the correct tool: what I'm generating isn't a display string, it's C# code.

Generating things such as `class Implementation_1_ISomeApi<@params>` : global::Their.Namespace.ISomeApi<@params>` and `typeof(Implementation_1_ISomeApi<>)` also needed some extra string manipulation.

Generating safe code, such as fully-qualifying all type names and escaping keywords, was also an opt-in choice, rather than the default. I'm concerned that many community-authored source generators simply won't take this level of care.

I'm also not sure how much harder things would have been if I hadn't have been able to use the `ISymbol`s from the interface I was implementing: I've got no way to generate my own symbols on which I can call `ToDisplayString()`.

I'm not sure what the best solution is here. I know there has been talk of adding a tool which helps users generate valid (and suitably qualified and escaped) C# code, but I don't have a clear picture of what I'd like such a tool to look like. Hopefully my use-cases can help.

## Unique names

When generating types using S.R.E., we have the luxury of unspeakable names, ensuring that field/local names don't clash with field/local names that the user has specified. Source generators don't have this. 

I ended up iterating through all members looking for potential collisions with my chosen field/local name, and appending numbers until I found something that was unique. I'm not sure what a good general solution is here: given that we (presumably) want source-generated types to be normal, valid C#, I don't think there's anything else which can be done.

This is another area where I fear that community source generator authors might not be so careful. Some sort of tool to help ensure that a name is unique in a given scope would be nice. Perhaps you give it a prefix, and it returns something starting with that prefix which is unique, both among members and among other generated names.

## Diagnostics

Being able to generate diagnostics was nice.

I did end up adding some of my own logic which ensured that I didn't output a `SourceText` if I also generated any error-level diagnostics, for a given type: if I hit errors when processing a user's interface, I wanted to keep processing it in case I discovered other errors that I wanted to report to the user, but I didn't want to output any generated implementation as that would certainly be invalid, and add further confounding errors. After all, errors mean that I *can't* generate a valid implementation.

Of course, there's no relationship between a `SourceText` and the diagnostics which were encountered when generating it in the current source generator API, so I had to implement this myself. Bear in mind that I did also generate compilation-level diagnostics, which don't refer to any specific user interface that I'm processing.

I don't know whether my case is common enough to consider adding explicit support for: I suspect it might be, as I suspect most source generators will be working on individual user-defined inputs (be they types or additional files), but I've only got one data point.

I did also discover `WellKnownDiagnosticTags.NotConfigurable`. This is useful, as any errors I encounter mean that the compilation process wasn't able to succeed, and I don't want the user suppressing those and then hitting errors at runtime instead. However I suspect many source generator authors might not be aware of it.

## Compiler versions

I'm assuming the user can have free choice over `LangVersion`, and this also affects code generated by source generators. Yes, source generators can look at the compilation to see what the langauge version is, but at the end of the day most will be written for some minimum C# language version, and won't be capable of generating anything below that.

I ended up picking C# 3 as my minimum language version, and generating C# source which used only C# 3 syntax.

I don't know whether it would be possible to allow source-generated files to use a different language version to the rest of the compilation: I suspect not. Even so, this would only move the problem: I could assume that the language version current when source generators were first released is available, but nothing beyond that.

Perhaps a good solution is to force source generators to declare the C# version of source that they generate. That at least forces authors to think about what they want to support, and means VS can give nice errors if the user has a lower `LangVersion` selected.

## Hiding generated types

I want my generated types to be hidden from intellisense. In recent VS versions, typing `ISomeApi` will also suggest my generated `Implementation_1_ISomeApi` type, even though it's in a different namespace. I don't want my generated type to clutter the user's intellisense, and I don't want the user to try and interact directly with it (they shoudld go through `RestClient.For<T>`).

`[EditorBrowsable(EditorBrowsableState.Never)]` doesn't affect types in the current assembly, and `[CompilerGenerated]` doesn't affect anything here. I did end up adding `[Obsolete]` onto my types's constructor so at least they can't instantiate it.

I think allowing `EditorBrowsable` to hide types in the current assembly would help a lot here.

## NuGet

My generated C# files rely on the user referencing my "main" NuGet package, because they reference various types from that package. The generated source also has a minimum version of the main package that it requires, since some required types were only introduced when I started writing the source generator. The source generator itself doesn't reference anything.

I therefore wanted my source generator NuGet package to declare a dependency on my main package, so that anyone trying to install the source generator automatically got the necessary reference to a suitable version of the main package. 

I couldn't find a way to do this using directives in my `csproj`: there didn't seem to be a way to say "I have a dependency on version X of package Y, but I don't actually want a reference to it". I ended up falling back to writing my own `.nuspec`, which does support this.

I suspect I'm not the only one in this situation: most source generated code will require types from another assembly. Adding the necessary support to `csproj` might be nice.

**Comment by Meowzz95** [comment]
Nice, I’m trying to write a source generator too, may I know how did you get a initial syntax tree and modify it? Do you write source as text then parse it?

**Comment by CyrusNajmabadi** [comment]
> I'm not sure what the best solution is here. I know there has been talk of adding a tool which helps users generate valid (and suitably qualified and escaped) C# code, but I don't have a clear picture of what I'd like such a tool to look like. Hopefully my use-cases can help.

I feel like this is out of scope for SourceGenerators.  i.e. this is a place for other libraries to step in and help (perhaps a 1st party ms one).  But hte point is to make these just strings so that you are not locked into a particular API.  I think there are spaces for syntactic APIs, semantic APIs, builder-based APIs, fluent APIs, low-level-highly-configurable APIs, high-level-very-opinionated-APIs etc. etc. etc.

I'm even working on one myself over here: https://github.com/dotnet/roslyn/pull/44124

The intent of mine is to be very opinionated and simple, and to not provide much (if any) control on the output.  

> Unique names

This should be really simple in practice.  Just use a unique prefix-suffix.  i.e. `__yourGeneratorName_theRest`.  There is practically no chance or concern about collisions in such a system.

> Diagnostics

Why can't you emit diagnostics with a DiagnosticAnalyer?

> Compiler versions

I don't see a problem here.  Emit what you want based on your desired userbase size.  Since it's generated code, there will very rarely be a need to actually use some new language feature (esp.  sine async/await has been out so long.

If a user wants to use your library which targets async/await, and they're on a prior version of hte language, too bad.  They can upgrade or not use the library.  It's a non-goal for source-generators (the feature) to target every lang version.   It may be *your* goal, in which case just only use 3.0 level features.

> Hiding generated types

I don't like this at all.  The point of generators is that they're generating *user code*.  It's a way of generating the code that they would write themselves.  If you don't want to clutter intellisense, that's easy.  Just put in a namespace called MyGeneratorInternals or something.

> I therefore wanted my source generator NuGet package to declare a dependency on my main package, so that anyone trying to install the source generator automatically got the necessary reference to a suitable version of the main package.

This seems out of scope, and against the idea of this being a 'Source Generator'.  The idea here is that this isn't a 'change everything generator', it's specifically about source.  Users can then decide how best to do things like add nuget/dll references if they so desire.

**Comment by CyrusNajmabadi** [comment]
>  I didn't work with syntax trees at all. The advice I'd heard before starting was to emit C# as strings directly and leave it as that, so that's the approach I took.

I don't think we're giving specific advice here.  If building a string is nice and simple, then do that.  If using SyntaxTrees is more effective, then use that.  If you end up building up a library of helpers, then use those.  If someone builds a helper API here, then use that.  etc. etc.

**Comment by canton7** [comment]
> I'm even working on one myself over here: #44124

Thanks! What's the deal with referencing libraries from source generators? The stuff I've read suggests this isn't something you should do.

> Just put in a namespace called MyGeneratorInternals or something.

I do. VS still suggests types in namespaces which haven't been imported. It's a new feature that I love, but it does get in the way here. I could name my types something completely unrelated to the thing they're implementing, which might be the best way out, but that seems perhaps unnecessarily user-unfriendly.

> Why can't you emit diagnostics with a DiagnosticAnalyer?

I could, but is there a reason I'd do that over having the source generator emit them directly?

> This seems out of scope, and against the idea of this being a 'Source Generator'. The idea here is that this isn't a 'change everything generator', it's specifically about source. Users can then decide how best to do things like add nuget/dll references if they so desire.

I'm not sure I agree. The source I'm generating needs a specific version of an external library. Is it more friendly to produce some cryptic compiler errors until they figure out the problem, or let NuGet ensure that they're referencing?

I do just want to stress that I'm not specifically asking for any changes: what I've got works, and I'm happy with source generators as they are now and what they let me do. I'm not going to argue for specific changes, that's entirely in MS's court if you want to. [This blog post](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) asked for feedback, and I gave mine just in case someone found here it useful 😄

**Comment by CyrusNajmabadi** [comment]
> I could, but is there a reason I'd do that over having the source generator emit them directly?

Because it would be redundant given an analzyer?  i.e. we already have a system for this.  So i would just use that.

**Comment by CyrusNajmabadi** [comment]
> I do. VS still suggests types in namespaces which haven't been imported. It's a new feature that I love, but it does get in the way here. I could name my types something completely unrelated to the thing they're implementing, which might be the best way out, but that seems perhaps unnecessarily user-unfriendly.

Not really seeing a problem.  If users wants to call your code, so be it.  I don't get hte idea that you're both generating code for them, but they should not be allowed to see it.

**Comment by canton7** [comment]
> Because it would be redundant given an analzyer? i.e. we already have a system for this. So i would just use that.

I'm confused. Source generators can already report diagnostics just fine. I'm using a system that you already have.

**Comment by CyrusNajmabadi** [comment]
> Thanks! What's the deal with referencing libraries from source generators? The stuff I've read suggests this isn't something you should do.

What'st he issue there?  I would expect it would be the same as referencing libraries from an analyzer.

**Comment by CyrusNajmabadi** [comment]
I guess i'm not really seeing the issue with diagnostics.  Report an issue.  And let that be enough.  I think worrying so much about what users will do isn't really important.

**Comment by canton7** [comment]
> What'st he issue there? I would expect it would be the same as referencing libraries from an analyzer.

The advice I've seen is "Don't have external dependencies in your analysers". Haooy to hear that that's wrong!

**Comment by CyrusNajmabadi** [comment]
> I'm not sure I agree. The source I'm generating needs a specific version of an external library. Is it more friendly to produce some cryptic compiler errors until they figure out the problem,

Why would it be a cryptic error?  Why not just doc this for your generator.

**Comment by canton7** [comment]
> I guess i'm not really seeing the issue with diagnostics. Report an issue. And let that be enough. I think worrying so much about what users will do isn't really important.

I'm still confused. What issue do you think I should report? I'm not trying to report an issue...

**Comment by CyrusNajmabadi** [comment]
> and I gave mine just in case someone here found it useful 😄

Absolutely.  I just want to weigh this complexity versus the value provided.

**Comment by CyrusNajmabadi** [comment]
> I'm still confused. What issue do you think I should report? I'm not trying to report an issue...

I read it as you wanted a way to just issue a diagnostic and have that suppress your source from being emitted into the compilation.  I was thinking htat that doesn't seem necessary.  i.e. i would expect hte generator to just always trudge on.  You've given the user the diagnostic.  it may in fact be better for them to see what problem they end up with the generated code due to that diagnostic.

**Comment by canton7** [comment]
> I was thinking htat that doesn't seem necessary. i.e. i would expect hte generator to just always trudge on. You've given the user the diagnostic. it may in fact be better for them to see what problem they end up with the generated code due to that diagnostic

Yeah, I don't know how general my case is. Some of the errors I report mean that I can't generate valid C#, and I don't want my diagnostic to be lost in a flood of compiler errors from my invalid generated code. I don't think it's unreasonable behaviour for a failed source generator to take steps to avoid an additional avalanche of compiler errors. 

As I said a few times, I'm not reporting issues or requesting changes here. What I've got works. 

I certainly don't want to report these errors from a separate analyser. That would mean processing the user's interface twice, which adds pointless cost for no gain.

> Absolutely. I just want to weigh this complexity versus the value provided

You're of course welcome to do what you want, but I'm not going to defend my corner, as there's no corner to defend.

**Comment by canton7** [comment]
> Why would it be a cryptic error? Why not just doc this for your generator.

I can guarantee that some people won't read any docs, and will be confused by the fact that they're apparently referencing the right thing but still being told that a particular type or member doesn't exist. Sure you could just say "oh well", but since it's such an easy case to avoid...

**Comment by CyrusNajmabadi** [comment]
Won't your diagnostic say: as a reference to this nuget package?

**Comment by CyrusNajmabadi** [comment]
> Some of the errors I report mean that I can't generate valid C#, and I don't want my diagnostic to be lost in a flood of compiler errors from my invalid generated code.

In that case, why not do an upfront check and just not emit in that case?

**Comment by canton7** [comment]
You want me to check that the right version of my library is referenced in the source generator? That could work, but I'm not necessarily convinced it's any better.

> In that case, why not do an upfront check and just not emit in that case?

For technical reasons. It's easier to interleave the checking with the emission in a single pass. I suppress the emission at the end if there were errors. This works fine, and I'm not asking for any changes here to support me.

**Comment by canton7** [comment]
I'm not going to defend myself against you any more, @CyrusNajmabadi. It's so draining. I put time into this write-up in good faith, in case someone found it useful. If you don't, please just close it and let's be done. I'm not here to fight.

**Comment by CyrusNajmabadi** [comment]
> in case someone found it useful. If you don't, please just close it and let's be done. I'm not here to argue.

I wasn't trying to argue.  My questions were to drill into this area more to get your perspective and weigh out what you were asking for.  My preference on feedback is to drill in and get as much of an understanding as possible so as to come up with best outcomes.

**Comment by b-straub** [comment]
@CyrusNajmabadi 
> i.e. this is a place for other libraries to step in and help (perhaps a 1st party ms one)
> I'm even working on one myself over here: #44124

Do you have an example how the code generation would look like? For my SG [ObservableTests](https://github.com/b-straub/ObservableTests) I'm using T4-Templates. Using a text template to prepare the generated code feels very natural.

**Comment by CyrusNajmabadi** [comment]
> Do you have an example how the code generation would look like?

You can see the tests added in that branch to get a sense of what it looks like.  Importantly, how the generated code looks is a non-goal for the effort.  i.e. there is no customization or control here.  You say what you want semantically/symbolically, and we just spit out the equivalent source for that, with no regard for any niceties.  i.e. names are fully qualified.  Formatting just uses `.NormalizeWhitespace()`, etc.

It's intended for easy construction, with teh view that generated code is just that, generated.

**Comment by b-straub** [comment]
> It's intended for easy construction, with teh view that generated code is just that, generated.

I totally agree. Generated code doesn't need to look nice. Finally it will be matter of taste and use case what approach to choose. Your approach is close to the syntax tree, mine to the _normal_ way of coding.

**Comment by MarkPflug** [comment]
> What'st he issue there? I would expect it would be the same as referencing libraries from an analyzer.

I was tinkering with source generators today, and ran into a very confusing issue due to referencing a library from my generator. If the generator is completely self-contained, everything worked fine. However, as soon as I started using an external library, the VS intellisense no longer worked. The MSBuild compile (f5) would work just fine, but in the IDE none of my generated code members would show up. It took me a *very* long time to realize that it was the external library causing the failure.

To be clear, this is a library that is being used by my generator only, not by the consumer of the generator. Both my generator and the external library targeted netstandard2.0. Should that work? For some reason I vaguely remember VS components might require a different target framework, but I can't find any specific details anymore. I imagine this scenario will be very common, as people start to use external libraries (like a CodeDom API) to assist in generating the code, which can be a bit tedious with just a stringbuilder.

I was following the blog post by Philip Carter, which didn't seem to mention any such restrictions.
